### PR TITLE
modularize make into shim

### PR DESCRIPTION
**Pull request by calvinmetcalf**
waiting on version 3.1.0 of pbkdf2-compat

### CODE REVIEW

**Comment by ForbesLindesay** [comment]
What are we waiting for in pbkdf2-compat v3.1.0?  I can't see any pull requests or branches?

**Comment by ForbesLindesay** [comment]
Ah, sorry.  I see that it actually has a few changes that don't appear to have been published to npm.

**Comment by calvinmetcalf** [comment]
it isn't waiting on it anymore it was actually 3.0.1 it was waiting on

**Comment by ForbesLindesay** [comment]
In that case, this should be ready to merge :)

**Comment by calvinmetcalf** [comment]
actually I had one more thing I forgot to push

**Comment by calvinmetcalf** [comment]
alright, this also adds publicDecrypt, privateEncrypt was able to test it against a manually compiled copy of iojs not sure how to test long term though

**Comment by dcousens** [comment]
Can we get some tests for this?

**Comment by calvinmetcalf** [comment]
yeah I can put together some vectors

**Comment by calvinmetcalf** [comment]
put in all the test from node, which did reveal I need to be able to handle certs see crypto-browserify/parse-asn1#4

**Comment by calvinmetcalf** [comment]
@dcousens ready for review again

File: test/index.js [hunk 1]
```diff
@@ -100,8 +101,58 @@ function testRun(i) {
...
+  var decryptedBuffer = myCrypto.privateDecrypt(rsaKeyPem, encryptedBuffer);
+  t.equal(input, decryptedBuffer.toString());
+
+  var decryptedBufferWithPassword = myCrypto.privateDecrypt({
+    key: rsaKeyPemEncrypted,
+    passphrase: 'password'
+  }, encryptedBuffer);
+  t.equal(input, decryptedBufferWithPassword.toString());
+
+  // encryptedBuffer = myCrypto.publicEncrypt(certPem, bufferToEncrypt);
```
Comment (dcousens) [inline test/index.js]: why are these comments here?

File: test/index.js [hunk 2]
```diff
@@ -100,8 +101,58 @@ function testRun(i) {
...
+
+  // decryptedBuffer = myCrypto.privateDecrypt(keyPem, encryptedBuffer);
+  // t.equal(input, decryptedBuffer.toString());
+
+  encryptedBuffer = myCrypto.publicEncrypt(keyPem, bufferToEncrypt);
```
Comment (dcousens) [inline test/index.js]: can you please isolate these tests from each other?  It looks like there is RSA encryption going on here as well as symmetric?

File: test/index.js (hunk 1 above)
Comment (calvinmetcalf) [inline test/index.js]: this is a test from io.js but I haven't implemented certificates. so I hope to be able to uncomment that so we can pass all of the io.js tests.

File: test/index.js (hunk 2 above)
Comment (calvinmetcalf) [inline test/index.js]: this is pulled verbatim out of io.js (well I changed the name of crypto) so that's why they are not isolated.

Comment (dcousens) [inline test/index.js]: Well, their tests suck.

Comment (dcousens) [inline test/index.js]: Well, lets isolate our tests, and then you can merge back? :)

**Comment by dcousens** [comment]
ACK once the tests are isolated.

**Comment by calvinmetcalf** [comment]
what do you mean by isolate our tests, as in add some publicDecrypt privateEncrypt tests that are not the node ones or move them to separate files?

**Comment by calvinmetcalf** [comment]
ping @dcousens what do you mean by initialize ?

**Comment by calvinmetcalf** [comment]
so hey iojs 1.1.0 has shipped with privateEncrypt so updated the tests to put it in and to pull out the node tests into their own file also (in a separate commit thinking of you @dcousens) changed all the tabs to spaces in the test.js file.  Then I realized there was a typo in the package.json (Was pointing it to my home built iojs) so fixed that.

**Comment by calvinmetcalf** [comment]
we are testing against 

- node 0.10.0 which has a full shim for all 4 functions
- node 0.11.0/0.12.0 which has a partial shim for only the 2 newer methods
- iojs which needs no shim

this ended up catching a big in how the shim was defined in 0.10

File: .travis.yml [hunk 3]
```diff
@@ -1,3 +1,6 @@
 language: node_js
 node_js:
-  - "0.11"
\ No newline at end of file
+  - "0.10"
+  - "0.11"
+  - "0.12"
```
Comment (dcousens) [inline .travis.yml]: ACK

File: browser.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+exports.publicEncrypt = require('./publicEncrypt');
+exports.privateDecrypt = require('./privateDecrypt');
+
+exports.privateEncrypt = function privateEncrypt(key, buf) {
+  return exports.publicEncrypt(key, buf, true);
```
Comment (dcousens) [inline browser.js]: IMHO this just reads confusingly.

Perhaps the following would read clearer?

``` javascript
__encrypt = require('./publicEncrypt')
__decrypt = require('./privateDecrypt')

exports.publicDecrypt = function(key, buf) { return __decrypt(key, buf) }
exports.publicEncrypt = function(key, buf) { return __encrypt(key, buf) }

exports.privateDecrypt = function(key, buf) { return __decrypt(key, buf, true) }
exports.privateEncrypt = function(key, buf) { return __encrypt(key, buf, true) }
```

File: index.js [hunk 5]
```diff
@@ -1 +1,18 @@
...
+if (typeof crypto.privateEncrypt !== 'function') {
+  exports.privateEncrypt = require('./browser').privateEncrypt;
+} else {
+  exports.privateEncrypt = crypto.privateEncrypt;
+}
+
+if (typeof crypto.publicDecrypt !== 'function') {
+  exports.publicDecrypt = require('./browser').publicDecrypt;
+} else {
+  exports.publicDecrypt = crypto.publicDecrypt;
```
Comment (dcousens) [inline index.js]: ACK

**Comment by dcousens** [comment]
I'm going to ACK this, but I haven't verified the tests.

**Comment by dcousens** [comment]
Thanks for your work on this @calvinmetcalf

**Comment by calvinmetcalf** [comment]
2.0.0

**Pull request merged**